### DISCUSSION TITLE
How to configure the config for dynamic input size training

### DISCUSSION THREAD

**Discussion by IT-five** [open-mmlab/mmclassification#1193] [General]
The configuration in configs/base/datasets is as follows, because I want to dynamically input the size for training. However, after I commented on resize, I saw that the official document requires default parameters for img_size in the backbone configuration file, but it will be used during training. An error is reported:Trying to resize storge that is not resizable

# dataset settings
dataset_type = 'ImageNet'
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    # dict(type='Resize', size=224),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='ImageToTensor', keys=['img']),
    dict(type='ToTensor', keys=['gt_label']),
    dict(type='Collect', keys=['img', 'gt_label'])
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    # dict(type='Resize', size=224),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='ImageToTensor', keys=['img']),
    dict(type='Collect', keys=['img'])
]

**Comment by mzr1996** [comment]
To enable dynamic input size, you need to format the `img` to `DataContainer` in the pipeline. The example can be found in MMDetction
https://github.com/open-mmlab/mmdetection/blob/e71b499608e9c3ccd4211e7c815fa20eeedf18a2/mmdet/datasets/pipelines/formatting.py#L217-L231

As for the `Trying to resize storage that is not resizable` error, maybe it's because some labels are empty, Refer to https://github.com/eriklindernoren/PyTorch-YOLOv3/issues/237, and you can check your dataset.

PS: Even if you enabled the dynamic input size, the shapes of batch images still need to be the same. Therefore, we need to pad the batch images to the same size, that's what the `DataContainer` is for.

**Reply by IT-five** [reply]
Thank you for your careful answer! Can you give a specific example?I don’t understand how to modify the pipeline of mmcls according to mmedet.

**Reply by mzr1996** [reply]
For example, modify the `ImageToTensor` as below:
```python
@PIPELINES.register_module()
class ImageToTensor(object):

    def __init__(self, keys):
        self.keys = keys

    def __call__(self, results):
        from mmcv.parallel import DataContainer
        for key in self.keys:
            img = results[key]
            if len(img.shape) < 3:
                img = np.expand_dims(img, -1)
            img = to_tensor(img.transpose(2, 0, 1))
            results['img'] = DataContainer(img, stack=True)
        return results

    def __repr__(self):
        return self.__class__.__name__ + f'(keys={self.keys})'
```

**Reply by IT-five** [reply]
Thank you very much for your answer
